text/0000-default-and-expanded-rustc-errors.md

```diff
@@ -0,0 +1,290 @@
+* Be careful using “ascii art” and avoid unicode. Instead look for ways to show the information concisely that will work across the broadest number of terminals.  We expect IDEs to possibly allow for a more graphical error in the future.
+* Where possible, use labels on the source itself rather than sentence "notes" at the end.  
+* Keep filename:line easy to spot for people who use editors that let them click on errors
+
+### Header
+
+```
+error[E0499]: cannot borrow `foo.bar1` as mutable more than once at a time
+  --> src/test/compile-fail/borrowck/borrowck-borrow-from-owned-ptr.rs:29:22
+```
+
+The header still serves the original purpose of knowing: a) if it's a warning or error, b) the text of the warning/error, and c) the location of this warning/error.  We keep the error code, now a part of the error indicator, as a way to help improve search results. 
+
+### Line number column
+
+```
+   |
+28 |
+   |
+29 |
+   |
+30 |
+31 |
+   |
+```
+
+The line number column lets you know where the error is occurring in the file.  Because we only show lines that are of interest for the given error/warning, we elide lines if they are not annotated as part of the message (we currently use the heuristic to elide after one un-annotated line).  
+
+Inspired by Dybuk and Elm, the line numbers are separated with a 'wall', a separator formed from pipe('|') characters, to clearly distinguish what is a line number from what is source at a glance.
+
+As the wall also forms a way to visually separate distinct errors, we propose extending this concept to also support span-less notes and hints.  For example:
+
+```
+92 |         config.target_dir(&pkg)
+   |                           ^^^^ expected `core::workspace::Workspace`, found `core::package::Package`
+   = note: expected type `&core::workspace::Workspace<'_>`
+   = note:    found type `&core::package::Package`
+```
+### Source area
+
+```
+      let bar1 = &mut foo.bar1;
+                      -------- first mutable borrow occurs here
+      let _bar2 = &mut foo.bar1;
+                       ^^^^^^^^ second mutable borrow occurs here
+      *bar1;
+  }
+  - first borrow ends here
+```
+
+The source area shows the related source code for the error/warning.  The source is laid out in the order it appears in the source file, giving the user a way to map the message against the source they wrote.
+
+Key parts of the code are labeled with messages to help the user understand the message.  
+
+The primary label is the label associated with the main warning/error.  It explains the **what** of the compiler message.  By reading it, the user can begin to understand what the root cause of the error or warning is.  This label is colored to match the level of the message (yellow for warning, red for error) and uses the ^^^ underline.  
```

Comment:
Could the error label be standard weight text rather than bold? One problem I have with the new messages is all the bold text - it makes the really important stuff (which should be bold) harder to pick out and (in my console at least) the bold text is quite a bit harder to read than standard weight.